### ISSUE TITLE
Example of search_around_sky use?

### ISSUE THREAD

**Issue by collioud** [astropy/astropy#3359]
Hi,

I am new to Astropy and I am trying to use the search_around_sky function, described [here](http://astropy.readthedocs.org/en/latest/api/astropy.coordinates.SkyCoord.html?highlight=search_around_sky#astropy.coordinates.SkyCoord.search_around_sky)

I tried (unsuccessfully) with this piece of code:
```
from astropy.coordinates import SkyCoord
from astropy.coordinates import Angle
cat = SkyCoord([1, 2, 3], [-30, 45, 8], unit="deg")
target = SkyCoord('1.1 -30.1', unit="deg")
cat.search_around_sky(target, Angle('2d'))
```
which finally results in `TypeError: 'Longitude' object with a scalar value does not support indexing`.

So, what is the proper way to use this function?
Is it possible to add a working example somewhere in the doc?

Thank you!
Best regards,
Arnaud

**Comment by embray** [comment]
I got the same thing.  Sounds like maybe a bug?

**Comment by cdeil** [comment]
I think this is a bug, too. The error occurs here ... both coords are indexed into which doesn't work for scalar inputs (in your example `target` is a scalar):
```python
d2ds = coords1[idxs1].separation(coords2[idxs2])
```
https://github.com/astropy/astropy/blame/03ffab593bc35af8c7387d66e1c2f82e5960853d/astropy/coordinates/matching.py#L342

@collioud For now you can call `np.atlead_1d` on your inputs before calling `search_around_sky`:
```python
import numpy as np
cat = np.atleast_1d(cat)
target = np.atleast_1d(cat)
```
@eteq Is this the correct fix to add this at the very start of the function?

---------

Concerning a usage example ... there is actually an example in the docs here:
http://astropy.readthedocs.org/en/latest/coordinates/matchsep.html#matching-catalogs

But it can't be found with search:
http://astropy.readthedocs.org/en/latest/search.html?q=match_coordinates_sky
so I think it would be better to put it in the function docstring (or link to it)
http://astropy.readthedocs.org/en/latest/api/astropy.coordinates.match_coordinates_sky.html

**Comment by collioud** [comment]
Thanks for the quick answer!
Unfortunaltely, this does not seem to solve the issue, since the new objects are considered now as `numpy.ndarray`.

Ex1:
```
cat = np.atleast_1d(cat)
target = np.atleast_1d(target)

cat.search_around_sky(target, Angle('2d'))
```
--> Error raised: `AttributeError: 'numpy.ndarray' object has no attribute 'search_around_sky'`

Ex2:
```
#cat = np.atleast_1d(cat)
target = np.atleast_1d(target)

cat.search_around_sky(target, Angle('2d'))
```
--> Error raised: AttributeError: 'numpy.ndarray' object has no attribute 'transform_to'

**Comment by cdeil** [comment]
@collioud You're right ... I presumed `np.atleast_1d` would return a `SkyCoord`, but it returns a `numpy.ndarray` object array with individual `SkyCoord` objects inside ... it's not what you want here.

Maybe the simplest would be to change your code to explicitly create array `SkyCoord` until this bug with scalar `SkyCoord` is fixed:
```python
import numpy as np
from astropy.coordinates import SkyCoord
from astropy.coordinates import Angle
from astropy.coordinates import search_around_sky
cat = SkyCoord([1, 2, 3], [-30, 45, 8], unit="deg")
# This doesn't work ... `target.shape = ()`, i.e. a scalar
target = SkyCoord('1.1 -30.1', unit="deg")
# This does work ... `target.shape = (1,)`, i.e. a length 1 array
target = SkyCoord([1.1,], [-30.1,], unit='deg')
cat.search_around_sky(target, Angle('2d'))
```
should give
```python
(array([0]), array([0]), <Angle [ 0.13225897] deg>, <Quantity [ 0.00230835]>)
```

**Comment by cdeil** [comment]
PS: thanks to this issue I noticed that `search_around_sky` exists and now I'm using it in my own code. :-)

**Comment by collioud** [comment]
Thank you Christoph!
It is a really useful function for astronomers who are always trying to find targets inside catalogs.

As far as I understood, results are for the moment ordered by catalog index. Will it be possible (someday) to have a parameter to order by separation and/or distance?

**Comment by cdeil** [comment]
Returning indices is the most flexible approach ... it's impossible to cover all the things users might want to do with the match result in Astropy itself.

One option is to store your match results in a [table](http://astropy.readthedocs.org/en/latest/table/index.html) and then to to further operations (like e.g. sort by distance or whatever you like) on the table ... this is what I'm currently doing for my application:
https://github.com/gammapy/gammapy/pull/235/files#diff-c85f1ce44bd2c14ef64e29fd56911d22R18

In my opinion it might be useful for Astropy to grow some higher-level matching API that returns results objects (possibly a Table subclass or wrapping Table), but if someone here agrees, please file a separate feature request, this one should be about fixing the bug you've found.
(Of course it's OK to ask here or on the mailing list if you have further questions about how to implement something!)

**Comment by collioud** [comment]
Yes, I tried few things with tables, and it seems I have everything I need...
Thanks again for your help!

**Comment by eteq** [comment]
It would be pretty easy to change this so that a scalar `SkyCoord` gets treated as a length-1 array `SkyCoord`.  But I'm not sure we *want* to do that.  That is, to me this is not a bug, but rather the intended behavior.

My reasoning is that ``search_around_sky`` was intended more for situations where both the target *and* the catalog are arrays.  That's why the return values are supposed indecies into the two `SkyCoord` objects.  But if it's a scalar that's input, that doesn't make sense.  So code like:
```
itarg, icat, d2d, d3d = target.search_around_sky(cat)
do_something_with(target[itarg], cat[icat])
```
will fail if `target` is a scalar.

The idea is that if you've got a single object you want to search for in a catalog, you can just do this:
```
seps = target.separation(cat)
np.argmin(seps)
```
And that gives you the same information, but in a way that's more readable given that `target` is a scalar.


@collioud - Does what I said above make sense?  As @cdeil noted, there's a whole documentation page on this, but clearly it's not that easy to find.  I'm not in favor of lengthy examples like that in the docstring, because these methods are all intertwined (as my above description demonstrates).  But it would definitely make sense to *reference* the http://astropy.readthedocs.org/en/latest/coordinates/matchsep.html#matching-catalogs page in the docstring.  If we did that, do you think you would have been able to figure this out based on what's written there?

**Comment by eteq** [comment]
BTW, @collioud @cdeil -  this also works but is very similar to what @cdeil said before:
```
from astropy.coordinates import SkyCoord
from astropy.coordinates import Angle
cat = SkyCoord([1, 2, 3], [-30, 45, 8], unit="deg")
target = SkyCoord(['1.1 -30.1'], unit="deg") #this line is different only in that it's a length-1 list of strings instead of just a string
cat.search_around_sky(target, Angle('2d'))
```
As I mentioned above, it wouldn't be at all hard to change the implementation of `search_around_sky` to work just like this this even with the original case @collioud brought up, but my opinion is that it would be *more* rather than less confusing.  But I am open to being persuaded otherwise!

**Comment by collioud** [comment]
@eteq Yes, I understand. In that case, if it is an intended behavior, maybe the doc related to the search_around_sky (http://astropy.readthedocs.org/en/latest/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.search_around_sky) should be more specific about the expected input parameters.

I suggest: 

search_around_sky(searcharoundcoords, seplimit)
Searches for all coordinates in this object around a supplied set of points within a given on-sky separation.

Parameters:	
searcharoundcoords : <b><u>array of</u></b> SkyCoord or BaseCoordinateFrame
The <b><u>coordinates</u></b> to search around to try to find matching points in this SkyCoord.

The rest of the description may stay identical...

**Comment by collioud** [comment]
And maybe, if possible, adding an explicit error message returned by `search_around_sky()`, like `First parameter has to be an array`...

**Comment by cdeil** [comment]
@eteq - I hadn't considered that issue ... I think you're right that it's better to fail early (in the function that computes array indices) instead of giving the user back array indices in the scalar case where indexing won't work.

It's not obvious for users that they should do something completely different (`seps = target.separation(cat); np.argmin(seps)`) for scalar target or catalog ... I think it might be easier to tell them how to convert their scalar SkyCoord into an array (what I tried to do with `np.atlead_1d`) ... but either way it's a valid use case and should be mentioned.

I think the main thing to do here is to fail with an error the user will understand.
@eteq Agreed? A `ValueError`?

And there's several ways in which the docstring of [SkyCoord.search_around_sky](http://astropy.readthedocs.org/en/latest/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.search_around_sky) can be improved ... besides the points already mentioned above:
* it talks about `coords1` and `coords2` when describing the return values, but those parameters only exist for the function version [astropy.coordinates.search_around_sky](http://astropy.readthedocs.org/en/latest/api/astropy.coordinates.search_around_sky.html#astropy.coordinates.search_around_sky)
* it says "BaseCoordinateFrame or SkyCoord", but I think the difference is that SkyCoord has actual coordinate values and this function doesn't make sense for BaseCoordinateFrame objects?

@eteq If you don't have time for this I could take a shot at making a PR.

**Comment by eteq** [comment]
Agreed with @collioud and @cdeil that a more explicit error message is a good idea.

I also  agree with @collioud and @cdeil's docstring suggestions and have implemented at least some of it in #3449. Lets move the discussion there for feedback on the specifics.
